letsencrypt/account.py

```diff
@@ -1,233 +1,203 @@
+        self.key = key
         self.regr = regr
+        self.meta = self.Meta(
+            # pyrfc3339 drops microseconds, make sure __eq__ is sane
+            creation_dt=datetime.datetime.now(
+                tz=pytz.UTC).replace(microsecond=0),
+            creation_host=socket.getfqdn()) if meta is None else meta
+
+        self.id = hashlib.md5(  # pylint: disable=invalid-name
+            self.key.key.public_key().public_bytes(
+                encoding=serialization.Encoding.DER,
+                format=serialization.PublicFormat.SubjectPublicKeyInfo)
+        ).hexdigest()
+        # Implementation note: Email? Multiple accounts can have the
+        # same email address. Registration URI? Assigned by the
+        # server, not guaranteed to be stable over time, nor
+        # cannonical URI can be generated. ACME protocol doesn't allow
+        # account key (and thus its fingerprint) to be updated...
 
     @property
-    def uri(self):
-        """URI link for new registrations."""
-        if self.regr is not None:
-            return self.regr.uri
-        else:
-            return None
+    def slug(self):
+        """Short account identification string, useful for UI."""
+        return "{1}@{0} ({2})".format(pyrfc3339.generate(
+            self.meta.creation_dt), self.meta.creation_host, self.id[:4])
+
+    def __repr__(self):
+        return "<{0}({1})>".format(self.__class__.__name__, self.id)
+
+    def __eq__(self, other):
+        return (isinstance(other, self.__class__) and
+                self.key == other.key and self.regr == other.regr and
+                self.meta == other.meta)
+
+
+def report_new_account(acc, config):
+    """Informs the user about their new Let's Encrypt account."""
+    reporter = zope.component.queryUtility(interfaces.IReporter)
+    if reporter is None:
+        return
+    reporter.add_message(
+        "Your account credentials have been saved in your Let's Encrypt "
+        "configuration directory at {0}. You should make a secure backup "
+        "of this folder now. This configuration directory will also "
+        "contain certificates and private keys obtained by Let's Encrypt "
+        "so making regular backups of this folder is ideal.".format(
+            config.config_dir),
+        reporter.MEDIUM_PRIORITY, True)
+
+    assert acc.regr.body.recovery_token is not None
+    recovery_msg = ("If you lose your account credentials, you can recover "
+                    "them using the token \"{0}\". You must write that down "
+                    "and put it in a safe place.".format(
+                        acc.regr.body.recovery_token))
+    if acc.regr.body.emails:
+        recovery_msg += (" Another recovery method will be e-mails sent to "
+                         "{0}.".format(", ".join(acc.regr.body.emails)))
+    reporter.add_message(recovery_msg, reporter.HIGH_PRIORITY, True)
+
+
+class AccountMemoryStorage(interfaces.AccountStorage):
```

Comment:
I like the idea of the interface...though I don't see much use for the MemoryStorage implementation.  Unittests have to be written for the filebased system anyway and integration tests are already creating temporary directories.